src/gridfs/download.rs

```diff
@@ -139,16 +140,196 @@ impl GridFsBucket {
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl GridFsDownloadStream {
+    /// Gets the file `id` for the stream.
+    pub fn files_id(&self) -> &Bson {
+        &self.file.id
+    }
+}
+
+impl AsyncRead for GridFsDownloadStream {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<std::result::Result<usize, futures_util::io::Error>> {
+        let stream = self.get_mut();
+
+        let future = match &mut stream.state {
+            State::Idle(ref mut cursor) => {
+                let new_future = get_bytes(
+                    cursor.take().unwrap(),
+                    buf.len(),
+                    stream.file.length,
+                    stream.file.chunk_size,
+                )
+                .boxed();
+                stream.state.set_busy(new_future)
+            }
+            State::Busy(future) => future,
+            State::Done => return Poll::Ready(Ok(0)),
+        };
+
+        let (result, cursor) = match future.poll_unpin(cx) {
+            Poll::Ready(result) => result,
+            Poll::Pending => return Poll::Pending,
+        };
+
+        stream.state = if !cursor.is_exhausted() {
+            State::Idle(Box::new(Some(cursor)))
+        } else {
+            State::Done
+        };
+
+        match result {
+            Ok(bytes) => {
+                buf[..bytes.len()].copy_from_slice(&bytes);
+                Poll::Ready(Ok(bytes.len()))
+            }
+            Err(error) => Poll::Ready(Err(error.into())),
+        }
+    }
+}
+
+async fn get_bytes(
+    mut cursor: Cursor<Chunk<'static>>,
+    buf_size: usize,
+    expected_file_len: u64,
+    chunk_size: u32,
+) -> (Result<Vec<u8>>, Cursor<Chunk<'static>>) {
+    let bytes_result = get_bytes_inner(&mut cursor, buf_size, expected_file_len, chunk_size).await;
+    (bytes_result, cursor)
+}
+
+async fn get_bytes_inner(
+    cursor: &mut Cursor<Chunk<'static>>,
+    mut buf_size: usize,
+    expected_file_len: u64,
+    chunk_size: u32,
+) -> Result<Vec<u8>> {
+    let mut bytes = vec![];
+    let mut current_n = 0u32;
+    while cursor.advance().await? {
+        let chunk = cursor.deserialize_current()?;
+
+        if chunk.n != current_n {
+            return Err(ErrorKind::GridFs(GridFsErrorKind::MissingChunk { n: current_n }).into());
+        }
+
+        let expected_len = FilesCollectionDocument::expected_chunk_length_from_vals(
+            expected_file_len,
+            chunk_size,
+            current_n,
+        );
+        if chunk.data.bytes.len() != (expected_len as usize) {
+            return Err(ErrorKind::GridFs(GridFsErrorKind::WrongSizeChunk {
+                actual_size: chunk.data.bytes.len(),
+                expected_size: expected_len,
+            })
+            .into());
+        }
+
+        let bytes_to_write = std::cmp::min(chunk.data.bytes.len(), buf_size) as usize;
+        bytes.extend_from_slice(&chunk.data.bytes[..bytes_to_write]);
```

Comment:
I was trying to figure out a way not to need to copy these bytes into the intermediary `Vec<u8>` that gets returned from this method, but I don't think it's possible. The two ways to do this would be:

- Pass a mutable reference to the `buf` provided in `poll_read` into this method, and copy the bytes directly in here. This won't work, however, because the lifetime of `buf` is contained to one call to `poll_read`, whereas a reference here would require it to be for the lifetime of the `GridFsDownloadStream`.
- Return something like a `Vec<&[u8]>` that holds pointers to each chunk we deserialize in this method. This isn't feasible because the lifetime of `chunk` (returned from the call to `deserialize_current`) is contained within the loop.

Reply:
If I'm following the logic right, the buffer being allocated each time will always have capacity equal to the chunk size from the files collection document except in the following cases:

- the final chunk is slightly smaller than a full chunk
- the output buffer is full
 
So in either case, a reusable buffer of capacity equal to the chunk size should be sufficient since the required capacity will be at most chunk size, just in some cases the entire buffer won't be used.

Reply:
Just to clarify -- I'm referencing the buffer being created on line 251. This method reads however many chunks are necessary to fill up the buffer the user provided (note the loop over the cursor), so the `bytes` vec isn't of a fixed size relative to the file/stream. Were you thinking of an intermediary buffer for the chunks being read?

Reply:
Ohh sorry, I misunderstood how the cached future works. I see that it takes in the length of the input buffer and reads up to that amount of bytes. 

>Were you thinking of an intermediary buffer for the chunks being read?

Yeah, I was thinking that instead of one future for the whole file, we could create a new future for each chunk. That way, we can use a chunk-size intermediary buffer in each future instead of one the size of the user's buffer / the file.

Reply:
Yeah, I actually had an implementation like that originally. My concern with that design, however, is the additional overhead that would be required to read through an entire file, particularly when the file is large and the chunk size is small. Whereas the current design only requires a single future to be created upon a single call to one of the `AsyncWriteExt` methods (e.g. `read`, `read_exact` or `read_to_end`), creating a new future for each chunk could possibly require a ton of loops through the state machine when a user calls `read_exact` or `read_to_end` to read in the requested number of bytes or the entirety of the file, respectively. It could also require lots of allocations in the way of creating all of those intermediary futures, although that could be mitigated by using tokio's [`ReusableBoxFuture`](https://docs.rs/tokio-util/latest/tokio_util/sync/struct.ReusableBoxFuture.html).

That said, `AsyncReadExt::read` (which calls down directly to `poll_read`) only says it returns "some bytes", and the presence of `AsyncReadExt::read_exact` and `AsyncReadExt::read_to_end` implies that `read` doesn't necessarily do either of those things. So perhaps it makes sense to be more conservative with how much we're actually reading in this method and allow the `AsyncWriteExt` methods to handle satisfying the quantity of bytes the user is requesting. I'm not sure how other libraries/implementations choose to define what "some" means; I can do some investigation into this. This would also be technically configurable by the user in that they could adjust their chunk size if we switch over to your proposal, although that seems like more of an implementation detail than a user knob and would require foresight upon inserting the data.

Reply:
>That said, read (which calls down directly to poll_read) only says it returns ["some bytes"](https://docs.rs/futures/latest/futures/io/trait.AsyncReadExt.html#method.read), and the presence of read_exact and read_to_end implies that read doesn't necessarily do either of those things. 

Yeah, I think the "some" here basically means "whatever is available". For example, in the case of a TCP socket, this might entail blocking until bytes have been received and just returning those bytes. In our case, it might mean a single chunk, but it could just as well mean "up to the number of bytes in the provided buffer". 

> This value would also be technically configurable by the user if we switch over to your proposal in that they could adjust their chunk size, although that seems like more of an implementation detail than a user knob and would require foresight upon inserting the data.

This is a good point. I think you're right that we're okay to buffer up to the size of the user's input (or a little more to fit the whole chunk in our case), so we probably don't need to do the multiple future route. This is what `tokio` does for `File` it looks like.

That said, it has made me realize that we _do_ need to update the future logic to support partial reads without discarding the unread portions of the chunks. I think that right now, it currently does discard any bytes in a chunk that exceed the users input buffer size. This is a problem when doing repeated reads (such as in `read_to_end`) with a buffer size < file size. Instead, we'll want to cache the leftover portion of the chunk and pick it up next time `poll_read` is called.

Reply:
Hm, I would think the spec tests for this would catch this error though. Could you write a test that calls `read` with a buffer half the chunk size repeatedly until it returns 0? The collected bytes from the `read` invocations _should_ be missing half of each chunk if my understanding of the various `AsyncRead` traits is correct. If the test passes then I might be missing something.